Child helpers

alexey-tikhonov opened the pull request.

aplopez commented on src/util/child_io.c:
```diff
@@ -0,0 +1,355 @@
+                       uint8_t **_buf,
+                       ssize_t *_len)
+{
+    return _read_pipe_recv(req, mem_ctx, _buf, _len);
```
If there is a good reason for this, can they be explicitly inlined?

alexey-tikhonov replied:
I didn't touch 'io' code in this PR (only logically split files). Out of scope of this PR.

alexey-tikhonov commented on src/util/child_common.c:
```diff
@@ -479,7 +479,7 @@ void child_terminate(pid_t pid)
+static struct tevent_timer *activate_child_timeout_handler(TALLOC_CTX *mem_ctx,
                                                  void *handler_pvt_ctx,
                                                  struct tevent_context *ev,
                                                  tevent_timer_handler_t handler,
```
Fixed.

aplopez commented on src/responder/ssh/ssh_cert_to_ssh_key.c:
```diff
@@ -184,85 +167,26 @@ static errno_t cert_to_ssh_key_step(struct tevent_req *req)
-done:
+    ret = sss_child_start(state, state->ev, P11_CHILD_PATH,
+                          state->extra_args, false, state->logfile,
+                          0, /* ssh cares only about exit code, so no 'io' */
```
Shouldn't this comment be close to the last argument (`NULL`)?

alexey-tikhonov replied:
Well, LLM clearly copied this from 29a698b9fb44f1769ce4ad235babed2915e5dac5

What I meant there was "SSSD doesn't want/need to read stdout of this child process", so neither 'child_out_fd' nor '_io' are needed.

Do you think it's worth re-writing comment?

aplopez commented on src/util/child_common.c:
```diff
@@ -40,8 +40,28 @@ struct sss_child_ctx {
+static int pvt_watch_destructor(struct sss_child_ctx **watch)
+{
+    if ((watch != NULL) && (*watch != NULL)) {
+        (*watch)->cb = NULL;
```
If `watch` is being freed, why do we want to set its fields to `NULL`. I could understand it if the memory pointed by the fields was freed, but that's not happening.

aplopez replied:
I understood it. I hadn't noticed that pvt_watch points to the parent context.

aplopez commented on src/providers/krb5/krb5_child.c:
```diff
@@ -4108,7 +4108,7 @@ int main(int argc, const char *argv[])
         {CHILD_OPT_SSS_CREDS_PASSWORD, 0, POPT_ARG_NONE, &sss_creds_password,
          0, _("Use custom version of krb5_get_init_creds_password"), NULL},
-        {CHILD_OPT_CHAIN_ID, 0, POPT_ARG_LONG, &chain_id,
+        {"chain-id", 0, POPT_ARG_LONG, &chain_id,
```
Why? What's wrong with the macro?

alexey-tikhonov replied:
I've removed macro from 'krb5_auth.h' because krb5 provider code shouldn't touch it anymore. So I could either add this macro to 'krb5_child.c" or use "chain-id" directly. I opted for the latter for the sake of consistency.

In general I'd like to add "chain-id" to 'SSSD_DEBUG_OPTS' (as well as dumpable" and "backtrace"), but this PR was already large enough and I didn't want to side track it.

aplopez commented on src/providers/proxy/proxy_auth.c:
```diff
@@ -179,12 +179,11 @@ static struct tevent_req *proxy_child_init_send(TALLOC_CTX *mem_ctx,
             "%s/proxy_child -d %#.4x --debug-timestamps=%d "
-            "--debug-microseconds=%d --logger=%s --domain %s --id %d "
-            "--chain-id=%lu",
+            "--debug-microseconds=%d --logger=%s --domain %s --id %d",
```
Why removing the argument from the call command?

alexey-tikhonov replied:
/gemini Can you review one specific patch that introduced this change being discussed in this thread?

alexey-tikhonov replied:
/gemini But does code in `proxy_child_init_send()` (or in 'proxy_auth.c') use new helper `sss_child_start()`?

alexey-tikhonov replied:
Yeap, that's indeed a bug. Good catch, @aplopez.

alexey-tikhonov replied:
Fixed.

alexey-tikhonov commented on src/util/child_common.c:
```diff
@@ -90,10 +110,20 @@ int child_handler_setup(struct tevent_context *ev, int pid,
     }
 
+    if (pvt != NULL) {
+        child_ctx->pvt_watch = talloc_zero(pvt, struct sss_child_ctx *);
```
Correct.

The idea is to allow user of this API to destroy high level "state context" without the need to bother with manual destruction of "child context" (or to track it all).

aplopez reviewed (approved):
LGTM. Thank you.

sumit-bose commented on src/responder/ifp/ifp_users.c:
```diff
@@ -1148,10 +1146,10 @@ ifp_users_find_by_valid_cert_send(TALLOC_CTX *mem_ctx,
     ret = sss_child_start(state, state->ev, P11_CHILD_PATH,
                           state->extra_args, false, state->logfile,
-                          STDOUT_FILENO,
+                          0, /* ifp cares only about exit code, so no 'io' */
```
Hi,

just a nitpick, `0` is still a valid file descriptor, maybe it would be better to use `-1` here or something like a `NO_CHILD_IO` define of `-1` or set it explicitly to `-1` in `sss_child_start()` if `_io` is `NULL` to make sure it is not used accidentally down in the call path in upcoming refactorings?

bye,
Sumit

alexey-tikhonov replied:
I added one more patch to address this, please take a look.

alexey-tikhonov replied:
/gemini review patch https://github.com/SSSD/sssd/pull/7991/commits/e41244e9be3477114accfed59e2c2e320e47155b

alexey-tikhonov commented:
Pushed PR: https://github.com/SSSD/sssd/pull/7991

* `master`
    * b871b0cc2e0189d9af08a16605763a5ce2d0f6f2 - CHILD HELPERS: make sure 'child_out_fd' isn't used
    * 8bddb6a510c3c1a88e31d43c6b1c66709be53193 - Renamed 'child_common.c' to 'child_handlers.c'
    * 18cba6e7d19aae8d13f59495c330953543459d75 - Cosmetics: indentation fix
    * fd92f450c2a10937c16934c647b308236d5a0c92 - KRB5 PASSKEY PLUGIN: ensure space for NULL termination
    * 301dc67a1d1e906954831255a7299f7fb19fbb6d - CHILD HELPERS: check return code of `sss_fd_nonblocking()`
    * 9f2e11ca58cccd84b8183baa26c20b0286200822 - CHILD HELPERS: cosmetics around namings
    * e4adf7e4a906f1b2a61e2a177596cca78114262b - CHILD HELPERS: make `exec_child_ex()` private
    * 4cd3aac5a08c50d7a2b6a6efe2bf81c34e948e8d - CHILD HELPERS: make `child_handler_setup()` and `child_handler_destroy()` kind of "private".
    * a831c000377c77c0a9ebee8eaba93260ed0af69a - CHILD HELPERS: make `child_io_destructor()` private
    * db8a601d08e320332ae661132f89211bb894eb41 - Cosmetics around close-fd helpers
    * 5e2586bb3ee144832b1f2fec3499dbc1666ec1d2 - DYNDNS: make use of `sss_child_start()`
    * 0e8f8876efbb0dd8f140e7cebc65a4c8cc782633 - PAM:PASSKEY: make use of `sss_child_start()`
    * 9be8b15b24e7dcc947688600b523ab096c464231 - PAM:CERT: make use of `sss_child_start()`
    * 452f205e689e2744852d89160511dba36a96de66 - IPA_SELINUX: make use of `sss_child_start()`
    * bd6dfc2ccb5a5f079c97ff176f1457e697e8dd09 - CHILD HELPERS: handle '--chain-id' as a basic arg
    * 96a7fc75d748a19430b3d2c4000b3a1eb48c55df - AD GPO: handle stuck 'gpo_child'
    * 839a73099e13ede49293733db9f6587037be51b4 - CHILD HELPERS: extend `sss_child_start()`
    * e5f4348fb889dc932179d8cf46257aac9adc94c8 - SSH: refactor ssh_cert_to_ssh_key.c to use sss_child_start()
    * 9f2ac0886a1c8c258a5624564b7c88e5aa55701a - sss_child_start(): allow NULL output _io arg
    * 5b72f114380f3cd34c61c4647d2048ec5d8171fa - responder/ifp: use sss_child_start() for p11_child certificate validation
    * ce528fd55986ccc877a4c3dcee968f4ac8677c39 - AD GPO: make use of `sss_child_start()`
    * 9324feb8a7f6eb27830f4047ec46cf7436460592 - AD pw renewal: make use of `sss_child_start()`
    * afffec389abf835859cb996ff21ac7b41ee3c1c8 - CHILD HELPERS: make activate_child_timeout_handler() internal
    * 92f977a46764960abf2644473563674b9dffbc57 - IDP: make use of `sss_child_start()`
    * 3b435ce3cedddffa811f41563f61be1e12110be2 - Delete 'exec_child()'
    * 511b44b8e7d0f75119f169bcb8a4cf9831f1c38d - LDAP: make use of `sss_child_start()`
    * 47b544a37457943c012c970d7b5fb1547f5972c9 - KRB5: make use of `sss_child_start()`
    * 81b2f20417e1103c6e6bd2901ad305ad71a5460c - New `sss_child_start()` helper
    * 10ad5a0ca9f1ff91cb89bb5008fd1755a1885885 - Rename 'sss_child_ctx_old' -> 'sss_child_ctx'
    * 91c528ac6e280010747a220d7022eac4fd0292e2 - Helpers defined in 'child_utils.h' aren't really used in child processes.
    * 24e9f9b15bcfbb8d8cf87a9fd256258fed3e4d2a - UTILS: don't use shared 'IN_BUF_SIZE'
    * 03f526333fe5bbe7dbf8db8ab335a60dc3a500bd - UTILS: split child helpers code
    * c00c6e214e96ee8f9b4a714f579b8c91b3e47a89 - CHILD_COMMON: unify structs 'response' and 'io_buffer'
    * 0fb034b7d247e704ab7bb3dd959ddc06cf603f2d - 'libsss_cert' doesn't use 'libsss_child'
    * 03da01d74376438ffb15c55869e1178ff7f6b97f - libkrb5 passkey plugin doesn't use 'libsss_util.so'
    * f9b226b015e8e262c591f11c5befc65a7a9a05d3 - Moved define used by ldap_child only out
    * 78c140081631af2668f59a1361f4e9741c02eae5 - UTILS: moved code used only by 'monitor'
    * 6f448e1cc5cf7d642bc3b822d85587fbb9aac86c - UTILS: removed stray declaration
